cfep-09.md

```diff
@@ -0,0 +1,99 @@
+2. Once the change is merged, the bot picks it up and starts going through the graph in topo order. 
+3. A migration PR is issued for a node only if 
+    1. The node depends on the changed pinnings.
+    2. It has no dependencies that depend on the new pinnings and have not been migrated.
+    3. It is not currently being migrated by an earlier migration. Per @scopatz, this state is stored in `libcflib` with a REST API.
+4. Process 3 continues until we determine that the migration is sufficiently complete and the change is merged into the global pinnings file. 
+
+### Implementation Details:
+- We would need to associate a version independent of git hash with each migration that orders them in time and thus orders the pinning files in time. A pinnings file could be versioned simply with a comment inserted at the top of the YAML containing the timestamp that the change was merged or something. We could use seconds from the epoch!
+- For a given migration, we only use the changes up to and including the commit that has the relevant pinning updates, even if further updates are merged for future migrations. This forces a single global pinning file to be used for a given subgraph that is being migrated. This seems required for consistency.
+- We allow for a local pinnings file to be present in each feedstock that would override the global pinnings if the version is later. In other words, the more recent pinnings file is always preferred by smithy. 
+- Smithy should use the most recent pinnings file when building the `.ci_support` files. This makes sure that rerendering the feedstock always produces the correct pinnings.
+- The migration bot would write the new pinnings file as the local one, rerender the repo and then issue the PR.
+- Migration changes have to be merged into the global pinnings file in the order they were made.
+
+## Rationale
+
+### Multiple migrations of the same package:
+We don't want to allow more than one change to the pinnings of a given node at a time, mostly because we cannot guarantee that later pinning migrations can be mixed with earlier ones (e.g., both could touch the same pins for some weird reason). We need to ensure that all nodes in the subgraph that share the same pinned dependence are uniformly migrated from one version to the next. However, we can have more than one migration in flight at a time simply because they could touch different parts of the graph. 
+
+### The rule of preferring the most recent pinnings file covers all of the relevant cases
+ - If a feedstock has no local pinnings, then that is fine. 
+ - If the file is present, but matches the global pinnings, this is also fine.
+ - In the middle of a migration, a feedstock that is being migrated would have a local pinnings file that should be used ahead of the global one. The local version would be ahead of global in this case since the migration bot would add the new pinnings file as the local one.
```

Comment:
What if we merge a new pinning before the current migration is finished? The global pinning will be later than the local one so it will be overridden even if the newest pinning needs a topo migration.

Reply:
I am not fully following the question here. A few things can happen and some details. 

1. New pinnings won't be merged to the global pinnings by hand. Only after we have determined a migration is "complete", would we merge them. In this case, the strict version ordering of merging the new pinnings to the global pinnings will ensure that the correct pinnings file is used for later migrations that are in progress. In other words, by strictly ordering the changes to the global pinnings and not letting migrations jump the line in merging, we ensure that the global pinnings file and its version is always before any current migration.

2. There is the case of us merging a migration when not all feedstocks are done. In that case, the global pinnings could eventually be one or several versions ahead of the local pinnings in the unmerged migration PR. A rerender of this PR would wipe out any local pinnings. Further any maintenance on unmigrated feedstocks or any new recipes that depend on the migrated feedstocks would then get the latest pinnings and thus be migrated out of topological order.

I should add a few comments on the rationale for case 2. (I should have added these  to the markdown, sorry for not including them!)

1. I consider this a "feature", not a bug. We have to move the entire ecosystem forward at a regular cadence even if some packages fall behind. We need the least worst solution here.

2. We don't have a formal policy for deprecating feedstocks at the moment. We do archive particularly out of date feedstocks, but we leave the packages in place on anaconda.org as opposed to marking them deprecated/broken. We could handle this situation with another CFEP on regularly versioning the main label for the conda-forge channels, but that is out of scope for this CFEP. Given this, simply leaving the unmigrated packages in the main channel is consistent with what we do now and handling them properly seems out of scope.

2. This procedure is consistent with how we are currently handling the great compiler migration. Right now if someone submits a new recipe, it will be built with the new compilers even if it has a dependence that has not yet ben migrated to them. The same thing can happen to an existing feedstock if it is rerendered as part of a version bump done by hand or some other maintenance.

Reply:
The final bit here is that the assumption in all of this is that migrating a pin will produce a package that is incompatible with the unmigrated pins due to the version constraint. Thus the conda solver should do the right thing even if we force a migration to end. If a person installs an unmigrated package, then the env gets rolled back.